src/qinfer/hyper_heuristic_optimisers.py

```diff
@@ -0,0 +1,202 @@
+    
+    def __call__(self,
+        n_pso_iterations=50,
+        n_pso_particles=60,
+        initial_position_distribution=None,
+        initial_velocity_distribution=None,
+        omega_v=0.35, 
+        phi_p=0.25, 
+        phi_g=0.5,
+        serial_map=map
+        ):
+        self._fitness_dt = np.dtype([
+            ('params', np.float64, (self._n_free_params,)),
+            ('velocities', np.float64, (self._n_free_params,)),
+            ('fitness', np.float64)])
+        self._fitness = np.empty([n_pso_iterations, n_pso_particles], dtype=self._fitness_dt)
+        local_attractors = np.empty([n_pso_particles], dtype=self._fitness_dt)
+        global_attractor = np.empty([1], dtype=self._fitness_dt)
+
+        if initial_position_distribution is None:
+            initial_position_distribution = distributions.UniformDistribution(np.array([[ 0, 1]] * self._n_free_params));
+            
+        if initial_velocity_distribution is None:
+            initial_velocity_distribution = distributions.UniformDistribution(np.array([[-1, 1]] * self._n_free_params))
+        
+        # Initial particle positions
+        self._fitness[0]["params"] = initial_position_distribution.sample(n_pso_particles)
+            
+        # Apply the boundary conditions if any exist
+        if self._boundary_map is not None:
+            self._fitness[itr]["params"] = self._boundary_map(self._fitness[itr]["params"])
+
+        # Calculate the initial particle fitnesses
+        self._fitness[0]["fitness"] = self.evaluate_fitness(self._fitness[0]["params"], 
+                                                            serial_map=serial_map)
+
+        # Calculate the positions of the attractors
+        local_attractors = self._fitness[0]
+        local_attractors, global_attractor = self.update_attractors(
+                                                self._fitness[0], 
+                                                local_attractors, 
+                                                global_attractor)
+
+        # Initial particle velocities
+        self._fitness[0]["velocities"] = initial_velocity_distribution.sample(n_pso_particles)
+        self._fitness[0]["velocities"] = self.update_velocities(
+                                                self._fitness[0]["params"], 
+                                                self._fitness[0]["velocities"], 
+                                                local_attractors["params"],
+                                                global_attractor["params"],
+                                                omega_v, phi_p, phi_g)
+
+        for itr in range(1, n_pso_iterations):
+            #Update the particle positions
+            self._fitness[itr]["params"] = self.update_positions(
+                self._fitness[itr - 1]["params"], 
+                self._fitness[itr - 1]["velocities"])
+
+            # Apply the boundary conditions if any exist
+            if self._boundary_map is not None:
```

Comment:
The behavior of ``_boundary_map`` seems to be similar to ``Model.canonicalize`` rather than ``Models.are_models_valid``? If so, that's quite reasonable, but should be clearly indicated in the documentation since it's an unusual way to specify boundaries.